AOC [23.01](https://adventofcode.com/2023/day/1).

Puzzle 1:

"Calibration document" has been altered by artist; can't read values on the document. 

Each line contains numbers. Need to find the first and last number in each line and combine them to form a number; these need to be added to find the calbration number.

Extracting numbers from a string of text is easy with a regular expression. I just need to do a little googling to remember the correct format!

I will look at the example and get it working, then use the actual data.

In [3]:
# Example
data = '''1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet'''

for line in data.split('\n'):
    print(line)

1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet


In [5]:
# Let's figure out the correct RegEx to get the numbers out of the text
# https://www.geeksforgeeks.org/python-regex/
import re
pattern = re.compile(r'\d')

for line in data.split('\n'):
    nums = re.findall(pattern,line)
    print(nums)




['1', '2']
['3', '8']
['1', '2', '3', '4', '5']
['7']


In [6]:
# Got the numbers; now need to combine them
for line in data.split('\n'):
    nums = re.findall(pattern,line)
    if len(nums) >= 2:
        number = int(f"{nums[0]}{nums[-1]}")
    else:
        number = int(f"{nums[0]}{nums[0]}")
    print(number)


12
38
15
77


In [7]:
# OK, all that works; let's run it on the actual data
input = '/home/alp/Google Drive/Python/AdventCode/2023/2301input.txt'

with open (input) as f_in:
    data = f_in.read().split('\n')

calibration_value = 0

for line in data:
    nums = re.findall(pattern,line)
    if len(nums) >= 2:
        number = int(f"{nums[0]}{nums[-1]}")
    else:
        number = int(f"{nums[0]}{nums[0]}")
    calibration_value += number

print(f"calibration value is {calibration_value}")

calibration value is 54561


Puzzle 2:

Some digits are spelled out with letters ('one','two',etc).

Need to find *real* first and last numbers on each line.

Solutions:
1. modify the RegEx pattern and use the same program
1. if that doesn't work, use the replace function on each string to swap out the written numbers for integers and then reuse the exact program above.

I'll mess around a bit with the first option but the second is pretty straight-forward.

In [51]:
#example
data = '''two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen'''

In [ ]:
pattern = re.compile(r'\d|')

for line in data.split('\n'):
    nums = re.findall(pattern,line)
    print(nums)

OK, that didn't take long -- it's going to be easier to do the replace than figure out the RE!

In [20]:
dd = {'one':'1','two':'2','three':'3','four':'4','five':'5','six':'6',
'seven':'7','eight':'8','nine':'9'}

for line in data.split('\n'):
    for k,v in dd.items():
        line = line.replace(k,str(v))
    print(line)
    
    nums = re.findall(pattern,line)
    if len(nums) >= 2:
        number = int(f"{nums[0]}{nums[-1]}")
    else:
        number = int(f"{nums[0]}{nums[0]}")
    print(number)


219
29
eigh23
23
abc123xyz
13
xtw134
14
49872
42
z1ight234
14
7pqrst6teen
76


This works, except when two numbers share the same first and last characters

```
eightwo = eight two
twone = two one
```
When this happens, if the k/v loop replaces a lower-value number (like one or two above), it destroys the number that appears earlier in the line (eight, two).

Let's go back to RegEx. . .

In [24]:
pattern = re.compile(r'\d|(one)|(two)|(three)|(four)|(five)|(six)|(seven)|(eight)|(nine)')

for line in data.split('\n'):
    nums = re.search(pattern,line)
    print(nums)

<re.Match object; span=(0, 3), match='two'>
<re.Match object; span=(0, 5), match='eight'>
<re.Match object; span=(3, 6), match='one'>
<re.Match object; span=(1, 4), match='two'>
<re.Match object; span=(0, 1), match='4'>
<re.Match object; span=(1, 4), match='one'>
<re.Match object; span=(0, 1), match='7'>


In [29]:
# That finds the first number. My earlier code finds the second number correctly.
findFirst = re.compile(r'\d|(one)|(two)|(three)|(four)|(five)|(six)|(seven)|(eight)|(nine)')

for line in data.split('\n'):
    num = re.search(findFirst,line)[0]
    if len(num) == 1:
        firstNum = int(num)
    else:
        firstNum = int(dd[num])
    print(firstNum)




2
8
1
2
4
1
7


In [52]:
dd = {'one':'1','two':'2','three':'3','four':'4','five':'5','six':'6',
'seven':'7','eight':'8','nine':'9'}

findFirst = re.compile(r'\d|(one)|(two)|(three)|(four)|(five)|(six)|(seven)|(eight)|(nine)')
findSecond = pattern = re.compile(r'\d')

for line in data.split('\n'):
    num = re.search(findFirst,line)[0]
    if len(num) == 1:
        firstNum = int(num)
    else:
        firstNum = int(dd[num])

    for k,v in dd.items():
        line = line.replace(k,str(v))
    secondNum = re.findall(findSecond,line)[-1]

    print(f"{firstNum}{secondNum}")







29
83
13
24
42
14
76


Finally -- got that to work! Now to run on the real data...

In [53]:
with open (input) as f_in:
    data = f_in.read().split('\n')

dd = {'one':'1','two':'2','three':'3','four':'4','five':'5','six':'6',
'seven':'7','eight':'8','nine':'9'}

findFirst = re.compile(r'\d|(one)|(two)|(three)|(four)|(five)|(six)|(seven)|(eight)|(nine)')
findSecond = pattern = re.compile(r'\d')

calibration_value = 0

for line in data:
    #print(line)
    num = re.search(findFirst,line)[0]
    if len(num) == 1:
        firstNum = int(num)
    else:
        firstNum = int(dd[num])

    for k,v in dd.items():
        line = line.replace(k,str(v))
    secondNum = re.findall(findSecond,line)[-1]

    calibration_value += int(f"{firstNum}{secondNum}")
    #print(f"{firstNum}{secondNum}")

print(f"calibration value is {calibration_value}")

calibration value is 54020


AOC says that 54020 is not the correct answer -- it's too low.

Let's try using the same method to look for the last number in the line -- just reverse the line!

In [54]:
print(f"{line} -- {line[::-1]}")

51cfsf6742cllbfnptkgttf -- fttgktpnfbllc2476fsfc15


In [55]:
with open (input) as f_in:
    data = f_in.read().split('\n')

dd = {'one':1,'two':2,'three':3,'four':4,'five':5,'six':6,'seven':7,'eight':8,'nine':9,
      'eno':1,'owt':2,'eerht':3,'ruof':4,'evif':5,'xis':6,'neves':7,'thgie':8,'enin':9}

findFirst =  re.compile(r'\d|(one)|(two)|(three)|(four)|(five)|(six)|(seven)|(eight)|(nine)')
findSecond = re.compile(r'\d|(eno)|(owt)|(eerht)|(ruof)|(evif)|(xis)|(neves)|(thgie)|(enin)')

calibration_value = 0

for line in data:
    num = re.search(findFirst,line)[0]
    if len(num) == 1:
        firstNum = int(num)
    else:
        firstNum = dd[num]

    num = re.search(findSecond,line[::-1])[0]
    if len(num) == 1:
        secondNum = int(num)
    else:
        secondNum = dd[num]

    calibration_value += int(f"{firstNum}{secondNum}")

print(f"calibration value is {calibration_value}")

calibration value is 54076
